File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/PostModelBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace CoolMvcBlog.Utils
{
    public class PostModelBinder : DefaultModelBinder
    {
        public override object BindModel(ControllerContext controllerContext, ModelBindingContext bindingContext)
        {
            if (controllerContext == null)
                throw new ArgumentNullException("controllerContext");
            if (bindingContext == null)
                throw new ArgumentNullException("bindingContext");

            if (bindingContext.Model == null)
            {
                var valueProviderResult = bindingContext.ValueProvider.GetValue("Id");

                if (!string.IsNullOrEmpty(valueProviderResult.AttemptedValue))
                {
                    int id = (int)valueProviderResult.ConvertTo(typeof(int));
                    bindingContext.ModelMetadata.Model = 
                        ObjectContextModule.CurrentContext
                        .PostSet.Include("Categories").Include("Author")
                        .Where(p => p.Id == id).Single();
                }
            }

            return base.BindModel(controllerContext, bindingContext);
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/PostModelBinder.vb
Public Class PostModelBinder
    Inherits DefaultModelBinder

    Public Overrides Function BindModel(ByVal controllerContext As ControllerContext, ByVal bindingContext As ModelBindingContext) As Object
        If controllerContext Is Nothing Then
            Throw New ArgumentNullException("controllerContext")
        End If

        If bindingContext Is Nothing Then
            Throw New ArgumentNullException("bindingContext")
        End If

        If bindingContext.Model Is Nothing Then
            Dim valueProviderResult = bindingContext.ValueProvider.GetValue("Id")

            If Not String.IsNullOrEmpty(valueProviderResult.AttemptedValue) Then
                Dim id = DirectCast(valueProviderResult.ConvertTo(GetType(Integer)), Integer)
                bindingContext.ModelMetadata.Model =
                    ObjectContextModule.CurrentContext.
                    PostSet.Include("Categories").Include("Author").
                    Where(Function(p) p.Id = id).Single

            End If
        End If

        Return MyBase.BindModel(controllerContext, bindingContext)
    End Function
End Class